UI/DesignWnd.cpp

```diff
@@ -676,53 +708,320 @@ namespace {
-            m_ordered_ids.push_back(id);
-            m_id_to_obsolete_and_loc[id] = std::make_pair(obsolete, --m_ordered_ids.end());;
+            m_ordered_design_ids.push_back(id);
+            m_id_to_obsolete_and_loc[id] = std::make_pair(obsolete, --m_ordered_design_ids.end());
         }
+
+        // Clear and load the ship hulls
+        m_hull_to_obsolete_and_loc.clear();
+        m_ordered_hulls.clear();
+        for (const auto& name_and_obsolete : hulls_and_obsoletes) {
+            const auto name = name_and_obsolete.first;
+            const auto obsolete = name_and_obsolete.second;
+            if (m_hull_to_obsolete_and_loc.count(name)) {
+                ErrorLogger() << "CurrentShipDesignManager::Load duplicate hull name = " << name;
+                continue;
+            }
+            m_ordered_hulls.push_back(name);
+            m_hull_to_obsolete_and_loc[name] = std::make_pair(obsolete, --m_ordered_hulls.end());
+        }
+
+        // Clear and load the ship parts
+        m_obsolete_parts = obsolete_parts;
     }
 
-    std::vector<std::pair<int, bool>> CurrentShipDesignManager::Save() const {
-        std::vector<std::pair<int, bool>> retval;
-        for (const auto id : m_ordered_ids) {
+    void CurrentShipDesignManager::Save(
+        std::vector<std::pair<int, boost::optional<bool>>>& design_ids_and_obsoletes,
+        std::vector<std::pair<std::string, bool>>& hulls_and_obsoletes,
+        std::unordered_set<std::string>& obsolete_parts)
+    {
+        design_ids_and_obsoletes.clear();
+        for (const auto id : m_ordered_design_ids) {
             try {
-                retval.push_back(std::make_pair(id, m_id_to_obsolete_and_loc.at(id).first));
+                design_ids_and_obsoletes.push_back(std::make_pair(id, m_id_to_obsolete_and_loc.at(id).first));
             } catch (const std::out_of_range&) {
                 ErrorLogger() << "CurrentShipDesignManager::Save missing id = " << id;
                 continue;
             }
         }
-        return retval;
+
+        hulls_and_obsoletes.clear();
+        for (const auto name : m_ordered_hulls) {
+            try {
+               hulls_and_obsoletes.push_back(std::make_pair(name, m_hull_to_obsolete_and_loc.at(name).first));
+            } catch (const std::out_of_range&) {
+                ErrorLogger() << "CurrentShipDesignManager::Save missing hull = " << name;
+                continue;
+            }
+        }
+
+        obsolete_parts = m_obsolete_parts;
+    }
+
+    //////////////////////////////////////////////////
+    //  AvailabilityManager                         //
+    //////////////////////////////////////////////////
+
+    /** A class to allow the storage of availabilty in a common location. */
+    class AvailabilityManager {
+        private:
+        // A tuple of bools for the state of the 3 availability filters:
+        // Obsolete, Available and Unavailable
+        std::tuple<bool, bool, bool> m_availabilities;
```

Comment:
Just having three bool members seems to me like it would be a lot simpler than dealing with the tuple in all the code below.